ios/RNSContentScrollViewDetector.mm

```diff
@@ -0,0 +1,101 @@
+#import "RNSContentScrollViewDetector.h"
+#import <React/RCTLog.h>
+#import <react/renderer/components/rnscreens/ComponentDescriptors.h>
+
+namespace react = facebook::react;
+
+@implementation RNSContentScrollViewDetector {
+  UIScrollView *_scrollView;
+  UIView *_linkedView;
+}
+
+- (void)didMoveToWindow
+{
+  if (self.window == nil) {
+    [self unregisterContentScrollViewInHierarchy];
+  } else {
+    if ((_scrollView = [self findContentScrollView])) {
+      [self registerContentScrollViewInHierarchy];
+    }
+  }
+}
+
+- (UIScrollView *)findContentScrollView
+{
+  auto scrollView = [RNSScrollViewFinder findScrollViewInFirstDescendantChainFrom:self];
+  if (scrollView) {
+    RCTLogInfo(@"Found content ScrollView");
+    return scrollView;
+  }
+
+  return nullptr;
+}
+
+- (void)registerContentScrollViewInHierarchy
+{
+  if (_linkedView != nullptr) {
+    RCTLogWarn(
+        @"Content ScrollView has already been registered. Make sure to only have one ScrollViewWrapper for content ScrollView.");
+    [self unregisterContentScrollViewInHierarchy];
+  }
+
+  UIView *parent = self.superview;
+  while (parent) {
+    if ([parent respondsToSelector:@selector(findContentScrollView)]) {
+      RCTLogWarn(
+          @"Nested ScrollViewWrapper detected. Make sure to only have one ScrollViewWrapper for content ScrollView.");
+    }
+
+    if ([parent respondsToSelector:@selector(registerContentScrollView:)]) {
+      [(id<ContentScrollViewConsumer>)parent registerContentScrollView:_scrollView];
+      _linkedView = parent;
+      RCTLogInfo(@"registered content ScrollView for: %@", parent);
+      break;
+    }
+
+    parent = parent.superview;
+  }
+}
+
+- (void)unregisterContentScrollViewInHierarchy
+{
+  if (_linkedView == nullptr) {
+    return;
+  }
```

Comment:
nit: it may take no action, so I'd say that it should be rather named `maybeUnregisterContentScrollViewInHierarchy`
tbh I'd consider other approaches here either:
1. do not calling this method from other places at all if `_linkedView` is null
2. having `maybeUnregisterContentScrollViewInHierarchy` which will call `unregisterContentScrollViewInHierarchy` which has the rest of logic

this one is very nitpicky, I'm okay with the current version, but I see potential improvement in terms of semantics, so I'm sharing some ideas